### PR TITLE
fix(apps-store): Fix exception on generating preview url for installed app screenshot

### PR DESCRIPTION
**Pull request by claucambra**
## Summary

Some installed apps meant for older server versions might unexpectedly offer up screenshot values in a non-string format (e.g. health). Avoid an exception by checking first if the first app screenshot is indeed a string and only then trying to create the $proxyPreviewUrl

Example error:

```
{"reqId":"HMJg7Rkp0hOomzjCuXPs","level":3,"time":"2024-10-27T11:40:53+00:00","remoteAddr":"192.168.21.3","user":"admin","app":"index","method":"GET","url":"/index.php/settings/apps/list","message":"OCA\\Settings\\Controller\\AppSettingsController::createProxyPreviewUrl(): Argument #1 ($url) must be of type string, array given, called in /var/www/html/apps/settings/lib/Controller/AppSettingsController.php on line 250 in file '/var/www/html/apps/settings/lib/Controller/AppSettingsController.php' line 236","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.0 Safari/605.1.15","version":"31.0.0.3","exception":{"Exception":"Exception","Message":"OCA\\Settings\\Controller\\AppSettingsController::createProxyPreviewUrl(): Argument #1 ($url) must be of type string, array given, called in /var/www/html/apps/settings/lib/Controller/AppSettingsController.php on line 250 in file '/var/www/html/apps/settings/lib/Controller/AppSettingsController.php' line 236","Code":0,"Trace":[{"file":"/var/www/html/lib/private/AppFramework/App.php","line":161,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Settings\\Controller\\AppSettingsController"},"listApps"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":306,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\Settings\\Controller\\AppSettingsController","listApps",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"settings.appsettings.listapps"}]},{"file":"/var/www/html/lib/base.php","line":1010,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/settings/apps/list"]},{"file":"/var/www/html/index.php","line":24,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","Line":146,"Previous":{"Exception":"TypeError","Message":"OCA\\Settings\\Controller\\AppSettingsController::createProxyPreviewUrl(): Argument #1 ($url) must be of type string, array given, called in /var/www/html/apps/settings/lib/Controller/AppSettingsController.php on line 250","Code":0,"Trace":[{"file":"/var/www/html/apps/settings/lib/Controller/AppSettingsController.php","line":250,"function":"createProxyPreviewUrl","class":"OCA\\Settings\\Controller\\AppSettingsController","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/apps/settings/lib/Controller/AppSettingsController.php","line":295,"function":"fetchApps","class":"OCA\\Settings\\Controller\\AppSettingsController","type":"->","args":[]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":208,"function":"listApps","class":"OCA\\Settings\\Controller\\AppSettingsController","type":"->","args":[]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":114,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Settings\\Controller\\AppSettingsController"},"listApps"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":161,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\Settings\\Controller\\AppSettingsController"},"listApps"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":306,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\Settings\\Controller\\AppSettingsController","listApps",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"settings.appsettings.listapps"}]},{"file":"/var/www/html/lib/base.php","line":1010,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/settings/apps/list"]},{"file":"/var/www/html/index.php","line":24,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/apps/settings/lib/Controller/AppSettingsController.php","Line":236},"message":"OCA\\Settings\\Controller\\AppSettingsController::createProxyPreviewUrl(): Argument #1 ($url) must be of type string, array given, called in /var/www/html/apps/settings/lib/Controller/AppSettingsController.php on line 250 in file '/var/www/html/apps/settings/lib/Controller/AppSettingsController.php' line 236","exception":{},"CustomMessage":"OCA\\Settings\\Controller\\AppSettingsController::createProxyPreviewUrl(): Argument #1 ($url) must be of type string, array given, called in /var/www/html/apps/settings/lib/Controller/AppSettingsController.php on line 250 in file '/var/www/html/apps/settings/lib/Controller/AppSettingsController.php' line 236"}}
```

## TODO

This is my first server PR so apologies if I messed something up

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [X] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [X] Screenshots before/after for front-end changes
- [X] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/settings/lib/Controller/AppSettingsController.php [hunk 1]
```diff
@@ -245,9 +245,11 @@ private function fetchApps() {
 		$apps = $appClass->listAllApps();
 		foreach ($apps as $app) {
 			$app['installed'] = true;
+			$appScreenshot = $app['screenshot'][0];
```
Comment (nickvergessen) [inline apps/settings/lib/Controller/AppSettingsController.php:248]: there are apps without screenshots, so the isset is needed …

Or we make it default to null now as there is a `is_string` check afterwards
```suggestion
			$appScreenshot = $app['screenshot'][0] ?? null;
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: apps/settings/lib/Controller/AppSettingsController.php (hunk 1 above)
Comment (claucambra) [reply apps/settings/lib/Controller/AppSettingsController.php]: Thanks for catching this, I took your suggestion

**Comment by nickvergessen** [comment]
I'll amend the actual fix...

**Review by nickvergessen — APPROVED** [review]

**Review by susnux — APPROVED** [review]
TIL: Do not trust the appinfo XSD.

**Comment by nickvergessen** [comment]
> TIL: Do not trust the appinfo XSD.

Well the problem is a bit that PHP's XML engine makes things a string until they appear multiple times :see_no_evil:

**Pull request merged**